apis/src/main/java/org/apache/rocketmq/apis/consumer/PullConsumer.java

```diff
@@ -0,0 +1,186 @@
+     * does not already have any metadata about the given topic.
+     *
+     * @param topic message's topic
+     * @return message queues of topic.
+     */
+    Collection<MessageQueue> topicMessageQueues(String topic) throws ClientException;
+
+    /**
+     * Manually assign messageQueue collections to this consumer.
+     * <p> Pull consumer must enable manual assignment mode in {@link PullConsumerBuilder} first, then call assign api.
+     * <p> This interface does not allow for incremental assignment and will replace the previous assignment.
+     * If the given collection is empty, it's treated same as unsubscribe().
+     * Manual assignment through this interface will disable the consumerGroup management functionality
+     * and there will be no rebalance operation triggered when group membership or cluster and topic metadata change.
+     * @param messageQueues are the collection for current consumer.
+     * @throws ClientException when assign
+     */
+    void assign(Collection<MessageQueue> messageQueues) throws ClientException;
+
+    /**
+     * Add subscription expression dynamically when use subscription mode.
+     *
+     * <p>If first {@link SubscriptionExpression} that contains topicA and tag1 is exists already in consumer, then
+     * second {@link SubscriptionExpression} which contains topicA and tag2, <strong>the result is that the second one
+     * replaces the first one instead of integrating them</strong>.
+     *
+     * @param subscriptionExpression new subscription expression to add.
+     * @return pull consumer instance.
+     */
+    PullConsumer subscribe(SubscriptionExpression subscriptionExpression) throws ClientException;
+
+    /**
+     * Remove subscription expression dynamically by topic.
+     *
+     * <p>Nothing occurs if the specified topic does not exist in subscription expressions of pull consumer.
+     *
+     * @param topic the topic to remove subscription.
+     * @return pull consumer instance.
+     */
+    PullConsumer unsubscribe(String topic) throws ClientException;
+
+    /**
+     * Get the collection of messageQueues currently assigned to current consumer.
+     * @return the collection of messageQueues currently assigned to current consumer
+     */
+    Collection<MessageQueue> assignments();
+
+    /**
+     * Pull messages for the topics specified by subscribe or assign api synchronously.
+     *
+     * <p> This method returns immediately if there are messages available.
+     * Otherwise, it will await the passed timeout. If the timeout expires, an empty map will be returned.
+     * An error occurs if you do not subscribe or assign messageQueues before pulling for data.
+     * @param maxMessageNum max message num when server returns.
+     * @return map of messageQueue to messageViews.
+     */
+    Map<MessageQueue, Collection<MessageView>> pull(int maxMessageNum) throws ClientException;
+
```

Comment:
sync pull can use pullAsync().get

Reply:
But it's not convenient to catch the real exception(ClientException) when using pullAsync.get.